### DISCUSSION TITLE
Latest update of hydration and virtual properties

### DISCUSSION THREAD

**Discussion by Dimonka2** [livewire/livewire#3305] [4. General]
What is a reason to have the following code in "HydratePublicProperties.php"?
````php
                if (version_compare(PHP_VERSION, '7.4', '<')) {
                    $instance->$property = $value;
                } else {
                    if((new ReflectionProperty($instance, $property))->getType()){ // <===========
                        is_null($value) || $instance->$property = $value;
                    } else {
                        $instance->$property = $value;
                    }
                }
````
This code crashes on a virtual properties that are managed by **__get** **__set**, `new ReflectionProperty()` causes an exception.

It brakes important logic of my components and giving no room to have virtual properties in a component.

**Comment by Dimonka2** [comment]
Just to give some background information. Virtual properties are required to solve issue of common query string for multiple components on a page based on the same logic like tables: pagination, search or filter query string.
There is discussion about pagination: https://github.com/livewire/livewire/discussions/3145
(still not resolved)

Possible solution that may avoid a need in virtual properties is aliases for query string: https://github.com/livewire/livewire/discussions/2496
(I cannot find my post with the same idea, maybe it was on a closed forum).

So this update is really making me disappointed, how to use Livewire after latest changes? 
Nobody is really using Livewire? Or maybe I am missing simple solution how to resolve query string collation of two components? 
Stop upgrading it?

**Comment by joshhanley** [comment]
@Dimonka2 this is the PR that introduced that line of code https://github.com/livewire/livewire/pull/3214

It didn't seem like it'd cause any problems, but it has caused your issue, which is an unintentional breaking change.

If you can submit PR that fixes your issue, but also supports what was intended in the PR, that would be welcome.

I would suggest locking your version to 2.5.1 in the meantime if it's causing a lot of issues.

I'm using Livewire heavily, but don't really use the query string, so can't offer any suggestions as to how to work around. Personally I try to minimise how many Livewire components are on a page for that and other reasons (like keeping state in sync between components).

Can you provide a small working example of your query string issue so I can see what you mean, and how you are using virtual properties?

Thanks!

**Reply by Dimonka2** [reply]
@joshhanley 
I am trying to minimize number of components on a page either. Refreshing takes a visible time.
But this issue is related to my package I use in all my projects that has a Livewire table component with pagination, search etc. Sometimes you need to put 2 tables like this on a page and of cause user expects that after refreshing a page they will stay on same page with the same filtering, number of elements in a table etc. For this is used query string. So I added a possibility to override query string element names by using virtual properties.
In order to make it working you need to define properties first via overriding `getPublicPropertiesDefinedBySubClass` function:
````php
    public function getPublicPropertiesDefinedBySubClass()
    {
        $data = parent::getPublicPropertiesDefinedBySubClass();
        $data[$this->pageQueryString] = $this->page;
        return $data;
    }
````
And then using __get __set to update actual properties.

It would be much easier to have a query string like `['page' => ['except' => 1, 'as' => $this->pageQueryString] ]` so that you do not need to create such overhead just to rename query string..

**Reply by joshhanley** [reply]
@Dimonka2 ah yeah makes sense. Yup some sort of renaming of the query string would be nice!

**Comment by ahinkle** [comment]
Hey there, I'm also having the same issue. Mine however seem to be specific to checkboxes being a `null` when they should actually be a boolean. Still trying to pinpoint it exactly. Were you able to find a solution?

![Screen Shot 2021-08-17 at 7 57 30 AM](https://user-images.githubusercontent.com/17038330/129730078-3d8bb71d-70d7-4b31-9ba8-eb2577fcd3eb.png)

```php
public $sameAsBilling = false;

public function save()
{
    dd($this->sameAsBilling); // returns null when it should return true.
}
```
